tuftool/src/update.rs

```diff
@@ -0,0 +1,142 @@
+use url::Url;
+
+#[derive(Debug, StructOpt)]
+pub(crate) struct UpdateArgs {
+    /// Follow symbolic links in `indir`
+    #[structopt(short = "f", long = "follow")]
+    follow: bool,
+
+    /// Number of target hashing threads to run (default: number of cores)
+    // No default is specified in structopt here. This is because rayon
+    // automatically spawns the same number of threads as cores when any
+    // of its parallel methods are called.
+    #[structopt(short = "j", long = "jobs")]
+    jobs: Option<NonZeroUsize>,
+
+    /// Key files to sign with
+    #[structopt(short = "k", long = "key", required = true, parse(try_from_str = parse_key_source))]
+    keys: Vec<Box<dyn KeySource>>,
+
+    /// Version of snapshot.json file
+    #[structopt(long = "snapshot-version")]
+    snapshot_version: NonZeroU64,
+    /// Expiration of snapshot.json file; can be in full RFC 3339 format, or something like 'in
+    /// 7 days'
+    #[structopt(long = "snapshot-expires", parse(try_from_str = parse_datetime))]
+    snapshot_expires: DateTime<Utc>,
+
+    /// Version of targets.json file
+    #[structopt(long = "targets-version")]
+    targets_version: NonZeroU64,
+    /// Expiration of targets.json file; can be in full RFC 3339 format, or something like 'in
+    /// 7 days'
+    #[structopt(long = "targets-expires", parse(try_from_str = parse_datetime))]
+    targets_expires: DateTime<Utc>,
+
+    /// Version of timestamp.json file
+    #[structopt(long = "timestamp-version")]
+    timestamp_version: NonZeroU64,
+    /// Expiration of timestamp.json file; can be in full RFC 3339 format, or something like 'in
+    /// 7 days'
+    #[structopt(long = "timestamp-expires", parse(try_from_str = parse_datetime))]
+    timestamp_expires: DateTime<Utc>,
+
+    /// Path to root.json file for the repository
+    #[structopt(short = "r", long = "root")]
+    root: PathBuf,
+
+    /// TUF repository metadata base URL
+    #[structopt(short = "m", long = "metadata-url")]
+    metadata_base_url: Url,
+
+    /// The working directory where the current metadata files will be written.
+    #[structopt(short = "w", long = "workdir", default_value = ".")]
+    workdir: PathBuf,
+
+    /// Directory of targets
+    indir: PathBuf,
+
+    /// Repository output directory
+    outdir: PathBuf,
+}
+
+impl UpdateArgs {
+    pub(crate) fn run(&self) -> Result<()> {
+        if let Some(jobs) = self.jobs {
+            rayon::ThreadPoolBuilder::new()
+                .num_threads(usize::from(jobs))
+                .build_global()
+                .context(error::InitializeThreadPool)?;
+        }
+
+        let new_targets = build_targets(&self.indir, self.follow)?;
+
+        let settings = tough::Settings {
+            root: File::open(&self.root).context(error::FileOpen { path: &self.root })?,
+            datastore: self.workdir.as_path(),
+            metadata_base_url: self.metadata_base_url.as_str(),
+            targets_base_url: self.metadata_base_url.as_str(),
```

Comment:
Are these intentionally the same?  I'm missing something.  Are we assuming "metadata" and "targets" are subdirectories of the given URL, the way they'd come out from the `create` command?  I don't understand how the repo loads successfully.

Reply:
In this case, since we aren't ever loading targets from the repo, we don't really care what the `targets_base_url` is. `tough::Settings` requires a value, so rather than asking the user for something we're never going to use, we use a placeholder of `metadata_base_url`.

I'll add a comment about this the clarify.